File: programmers/js/new/level3/N으로_표현.js
var answer = Number.MAX_SAFE_INTEGER;
function solution(N, number) {
    dfs(0, 0, N, number);
    if (answer === Number.MAX_SAFE_INTEGER) {
        answer = -1;
    }
    return answer;
}

function dfs(count, result, N, number) {
    if (count > 8) return;
    // NN
    let NN = "";
    for (let i = 1; i < 9 - count; ++i) {
        NN += N;
        const num = parseInt(NN);
        // N+N
        let res = result + num;
        if (res === number && count + i < answer) {
            answer = count + i;
        }
        dfs(count + i, res, N, number);
        // N-N
        if (result !== num) {
            res = result - num;
            if (res === number && count + i < answer) {
                answer = count + i;
            }
            dfs(count + i, res, N, number);
            res = num - result;
            if (res === number && count + i < answer) {
                answer = count + i;
            }
            dfs(count + i, res, N, number);
            // N*N
            res = result * num;
            if (res === number && count + i < answer) {
                answer = count + i;
            }
            dfs(count + i, res, N, number);
        }
        // N/N
        if (num !== 0) {
            res = parseInt(result / num);
            if (res === number && count + i < answer) {
                answer = count + i;
            }
            dfs(count + i, res, N, number);
        }
        if (result !== 0) {
            res = parseInt(num / result);
            if (res === number && count + i < answer) {
                answer = count + i;
            }
            dfs(count + i, res, N, number);
        }
    }
}

console.log(solution(5, 12));

File: programmers/python/level3/N으로_표현.py
answer = 9
ops = ['+', '-', '*', '/']
def solution(N, number):
    for op in ops:
        dfs(op, 0, number, N, 0)
    return -1 if answer == 9 else answer

def dfs(type, current, target, N, count):
    if current == target:
        global answer
        answer = min(answer, count)
        return
    if count > 8:
        return
    num_str = ''
    for i in range(1,9-count):
        num_str += str(N)
        num = int(num_str)
        if type == '+':
            next = current + num
        elif type == '-':
            next = current - num
        elif type == '*':
            next = current * num
        elif type == '/':
            next = current / num
        for op in ops:
            dfs(op, next, target, N, count+i)